### PR TITLE
Gary/feature/update delete expenses

### PR DESCRIPTION
**Pull request by garysmith1933**
This builds on top of my 2 other PRs. Please review them first before this one.

# Description
This commit implements the ability to update and delete existing expenses. 

## Related to the following Jira/Trello issue 

I will link it if I can find one after.

## Main changes explained:
- Created API to handle updating existing expenses.
- Created API to handle deletion of existing expenses.
- Created deleteFromS3 method in S3-Upload file to handle deletion of existing expense receipts.
- Moved S3-Upload file to Groups folder instead of its own thing.
- Added feature to allow contributions to be added and updated in addExpense and updateExpense apis.
- Added ability to update an expense as "paid".

## How to test:

Since this branch includes 2 main features.  Here are two different sets of instructions. I 
0. I'm sorry in advance. 
1. check into current branch
2. do `pnpm install` and `pnpm run dev` to run this PR locally
3. Make sure you are logged in.

NOTE: I added contributions to this branch. To test this add contributions as a key in the form data, the value should be an array of objects that will have this format. Ex: [{"member_id": "66f83db2e64fda3c2783ae10", "amount": 100}, {"member_id": "66f83e89dcdaaa757d8642b2", "amount": 50]}. 

I also added an is_paid field, this is just a simple way to allow an expense be updated as completed. the key for this in the form data is "is_paid" the value will need to be set to either true, or false.

Updating Expenses:
1. Create a new group.
2. Add at least 2 - 3 users to the newly created group via a post request.
3. Add an expense to the new group the URL of the request should look like localhost:3000/api/groups/ *insert group id here*/ addExpense - Make sure you are entering the expense data as form-data using key-value pairs, otherwise you will get an error.
5. Update the newly created expense, the URL of the request should look like localhost:3000/api/groups/*insert group id here*/expenses/*insert expense id here*/updateExpense. You want to make sure all fields are present with some of the values changed however you want. Verify the expense returned matches the new changes.
  
Deleting Expenses:
1. Delete the updated expense, the URL of the request should look like localhost:3000/api/groups/*insert group id here*/expenses/*insert expense id here*/deleteExpense. 
2. Check the response, and verify the message says that the expense is deleted.
3. Confirm that the expense is deleted by making sure that no UserExpenses have the Id of the expense you created. 
5. Confirm that the expense is deleted by making sure that the expense is no longer shown in the expenses array of the group.
6. Confirm that the expense is deleted by making that the expense is no longer present in the list of expenses.

## Note:
I thought this would be much less complex, hence why I had these both be on one branch. Sorry about that.
I also know the url of the requests can be quite long. If its a problem, I can try to shorten them, Just let me know!

### CODE REVIEW

**Review by cvtqx — COMMENTED** [review]
Updating Expenses works as described.

Deleting Expenses:

-Expense gets deleted from the group
 However I get this response:
Screenshot 2024-10-10 at 11 25 14 AM

When checking the database - I can still see the expense in userExpenses and in Expenses collections, but it is no longer shown in the expenses array of the group.

File: app/api/groups/route.ts [hunk 1]
```diff
@@ -27,12 +28,14 @@ export const POST = async(request: NextRequest): Promise<NextResponse> => {
...
     await dbConnect();
     const body: GroupBody = await request.json();
+    let inviteLink = nanoid(7);
 
     const group = await Group.create({
       name: body.name,
       description: body.description,
       budget: body.budget,
       admin_id: body.user_id,
+      invite_link: inviteLink
```
Comment (cvtqx) [inline app/api/groups/route.ts:38]: FYI - comma missing after inviteLink

**Comment by garysmith1933** [comment]
@cvtqx Can you try again with a newly created group? I tried it using the party animals one and didn't have a problem. 
Screenshot (176)

**Comment by cvtqx** [comment]
Tried testing this again, made sure I was logged in. Add and update expense work seamlessly, however delete expense - I get the same error when I try to delete an expense that I have just created.
Screenshot 2024-10-12 at 12 17 21 PM

however, if I try to delete a random expense from any group in the database, it works just fine and expense disappears from the database. Perhaps it has something to do with the way I am creating an expense? Should the 'contributions' array include the current user id? I am bad at debugging  and not sure where to add logs. I will try reviewing this again later.

**Review by cvtqx — APPROVED** [review]
Works! so when you add an expense, the contributions array MUST include every member of the group, then it works upon deletion.
Screenshot 2024-10-12 at 12 55 22 PM

**Pull request merged**